core/src/main/scala/chisel3/internal/firrtl/IR.scala

```diff
@@ -339,6 +339,25 @@ private[chisel3] object ir {
       extends Definition
   case class DefObject(sourceInfo: SourceInfo, id: HasId, className: String) extends Definition
 
+  class Placeholder(val sourceInfo: SourceInfo) extends Command {
+    private var buffer: mutable.Builder[Command, ArraySeq[Command]] = null
+
+    def getBuffer: mutable.Builder[Command, ArraySeq[Command]] = {
+      if (buffer == null)
+        buffer = ArraySeq.newBuilder[Command]
+      buffer
+    }
+
+  }
+
+  object Placeholder {
+    def unapply(placeholder: Placeholder): Option[(SourceInfo, Seq[Command])] = {
+      Some(
+        (placeholder.sourceInfo, placeholder.getBuffer.result())
+      )
+    }
+  }
```

Comment:
Do we need to check that we don't accidentally append to the buffer after calling result?
We probably also need to memoize the result since: `After calling result() the behavior of a Builder is undefined. No further methods should be called.` [[1]](https://www.scala-lang.org/api/2.13.16/scala/collection/mutable/Builder.html). It's not clear that you can even call `result()` multiple times.

Reply:
This is beefed up to work similarly to `Block` in https://github.com/chipsalliance/chisel/pull/4628/commits/1bc48a5828b60d417f6c8a01a87095d69ac5c10b. Thanks for the pointer to the docs. It's somewhat surprising that this is just undefined as opposed to being "read once" as I had thought.